Assign meta data

miaoyinb opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
MGs that depend on metadata cannot use metadata-free meshes (especially file mesh) as input.

## Design
Add simple-type metadata to existing mesh (especially file mesh).

## Impact
Expand the applications of MGs.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

miaoyinb committed:
Add Mesh Metadata to Existing Meshes

closes #23135

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
add Forward declarations

MengnanLi91 committed:
add Forward declarations

MengnanLi91 committed:
add Forward declarations

MengnanLi91 committed:
Merge #17897 to current version

MengnanLi91 committed:
Merge #17897 to current version

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

style change for Precheck test.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
style change for Precheck test.

MengnanLi91 committed:
style change for Precheck test.

MengnanLi91 committed:
Fix reset mechanism #22884

MengnanLi91 committed:
Fix reset mechanism #22884

MengnanLi91 committed:
Fix reset mechanism #22884

MengnanLi91 committed:
style changes for precheck test

MengnanLi91 committed:
style changes for precheck test

MengnanLi91 committed:
Add doco for solution invalid and add newsletter blurb

MengnanLi91 committed:
Add doco for solution invalid and add newsletter blurb

MengnanLi91 committed:
Work in progress on invalid solution

MengnanLi91 committed:
Work in progress on invalid solution

working on making test case

update test case(not done yet)

MengnanLi91 committed:
Work in progress on invalid solution

MengnanLi91 committed:
working on making test case

MengnanLi91 committed:
working on making test case

MengnanLi91 committed:
update test case(not done yet)

MengnanLi91 committed:
update test case(not done yet)

miaoyinb commented:
@GiudGiud
As I mentioned before. This is the draft update for mesh metadata assignment. Your comments would be appreciated so that we could improve this before the official PR. Thank you in advance.
(I have another one for metadata retaining)

MengnanLi91 committed:
Add SolutionInvalidityRegistry

MengnanLi91 committed:
Add SolutionInvalidityRegistry

MengnanLi91 committed:
Add SolutionInvalidityRegistry

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity descriptions

zachmprince committed:
Refactoring optimization reporters

zachmprince committed:
Adding mesh-based optimization parameters #22995

GiudGiud commented on framework/doc/content/source/meshgenerators/AddMetaDataGenerator.md:
```diff
@@ -0,0 +1,15 @@
+
+## Overview
+
+This MeshGenerator object allows the user to add extraneous Mesh Metadata to the input mesh. This would allow the input mesh, especially from `FileMeshGenerator`, to be compatible with those mesh generators (e.g., mesh generators in the Reactor Module) that depend on Mesh Metadata to function.
```
let s say external instead of extraneous. 
extraneous also means irrelevant

GiudGiud commented on framework/doc/content/source/meshgenerators/AddMetaDataGenerator.md:
```diff
@@ -0,0 +1,15 @@
+
+This MeshGenerator object allows the user to add extraneous Mesh Metadata to the input mesh. This would allow the input mesh, especially from `FileMeshGenerator`, to be compatible with those mesh generators (e.g., mesh generators in the Reactor Module) that depend on Mesh Metadata to function.
+
+## Further AddMetaDataGenerator Documentation
```
Probably should be an example of starting from a clean exodus

what do we have that uses metadata?
- drums
- trimmability
- pitches and sizes for making patterns
is there anything outside the reactor module?

miaoyinb replied:
It seems that only Reactor Module is using meta data. I can add a test case to one of them it that works for you.

GiudGiud commented on framework/include/meshgenerators/AddMetaDataGenerator.h:
```diff
@@ -0,0 +1,73 @@
+#include "MeshGenerator.h"
+
+/**
+ * This mesh generator assigns extraneous mesh metadata to the input mesh
```
same

GiudGiud commented on framework/src/meshgenerators/AddMetaDataGenerator.C:
```diff
@@ -0,0 +1,156 @@
+  params.addParam<std::vector<std::string>>("real_scalar_metadata_names",
+                                            "Names of the real scalar mesh metadata.");
+  params.addParam<std::vector<Real>>("real_scalar_metadata_values",
+                                     "Types of the real scalar mesh metadata.");
```
Values instead of types?

miaoyinb replied:
Sorry, it was from a previous version.

GiudGiud commented on framework/src/meshgenerators/AddMetaDataGenerator.C:
```diff
@@ -0,0 +1,156 @@
+    paramError(param_name + "_metadata_values",
+               "Length of this parameter must be the same as that of " + param_name +
+                   "_metadata_names");
+}
```
civet wont lack this missing newline

miaoyinb replied:
Still struggling with transition from Atom to VSCode. 😞

GiudGiud replied:
there's a vscode channel on slack btw where people share tips

GiudGiud reviewed (commented):
Thanks for the sneak peak!

I think we ll want bools as well?

miaoyinb commented:
> Thanks for the sneak peak!
> 
> I think we ll want bools as well?

There seems to be some issues about `std::vector<bool>` as an input parameter. Will look into this.

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672872012 -0700

parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672869399 -0700

parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672869270 -0700

parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672868931 -0700

Update summary table and add test

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
update output method

Update solution_invalid.i

MengnanLi91 committed:
update output method

MengnanLi91 committed:
Update solution_invalid.i

MengnanLi91 committed:
update format

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
update SolutionInvalidity

Update SolutionInvalidity.C

Update SolutionInvalidity.C

MengnanLi91 committed:
update SolutionInvalidity

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
Update SolutionInvalidity.C

MengnanLi91 committed:
Update RunParallel.py

MengnanLi91 committed:
Test console error

MengnanLi91 committed:
Test console error

MengnanLi91 committed:
update console printout

MengnanLi91 committed:
update console printout

loganharbour committed:
parent 16ecc007b4f053c58d817fb3d5b09dee3720c606
author Logan Harbour <loganhharbour@gmail.com> 1671745211 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672868105 -0700

parent 16ecc007b4f053c58d817fb3d5b09dee3720c606
author Logan Harbour <loganhharbour@gmail.com> 1671745211 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672867926 -0700

Get PerfGraph dataStore/Load working again

style change for Precheck test.

MengnanLi91 committed:
Update SolutionInvalidInterface.h

MengnanLi91 committed:
Update RunParallel.py

MengnanLi91 committed:
Update tests

Update tests

Update tests

lindsayad committed:
Add SRS, SDD, and RTM docs for modules I own

At least for the ones that have template pages:
- scalar transport
- fluid structure interaction
- heat conduction

Refs #15968

MengnanLi91 committed:
expand threshold from number to vector

miaoyinb commented:
@Gu

> > Thanks for the sneak peak!
> > I think we ll want bools as well?
> 
> There seems to be some issues about `std::vector<bool>` as an input parameter. Will look into this.

Done. The scalar boolean metadata is added. The vector boolean metadata is not added because they do not exist so far and `vector<vector<bool>>` input parameter is not support so far.

Let me know if any additional improvements are needed before the official PR. Thanks @GiudGiud

GiudGiud reviewed (commented):
so in the module we have
```
src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:    _quad_center_block_id(declareMeshProperty<subdomain_id_type>(
src/meshgenerators/PinMeshGenerator.C:    declareMeshProperty("duct_region_ids", std::vector<std::vector<subdomain_id_type>>());
src/meshgenerators/PatternedHexMeshGenerator.C:        declareMeshProperty<std::vector<Point>>("control_drum_positions", std::vector<Point>())),
```

should we have these types here too or is it going to be able to convert ?

GiudGiud commented on framework/include/meshgenerators/AddMetaDataGenerator.h:
```diff
@@ -0,0 +1,77 @@
+  /// List of mesh metadata values for the dof_id_type type scalars
+  const std::vector<dof_id_type> _dof_id_type_scalar_metadata_values;
+  /// List of mesh metadata names for the boolean type scalars
+  const std::vector<std::string> _boolean_scalar_metadata_names;
```
should be bool ?

MengnanLi91 committed:
fix moose test threads

MengnanLi91 committed:
back to number for testing

miaoyinb commented:
> so in the module we have
> 
> ```
> src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:    _quad_center_block_id(declareMeshProperty<subdomain_id_type>(
> src/meshgenerators/PinMeshGenerator.C:    declareMeshProperty("duct_region_ids", std::vector<std::vector<subdomain_id_type>>());
> src/meshgenerators/PatternedHexMeshGenerator.C:        declareMeshProperty<std::vector<Point>>("control_drum_positions", std::vector<Point>())),
> ```
> 
> should we have these types here too or is it going to be able to convert ?

Yes, these should also be added.

loganharbour committed:
Add mutex to setSolutionInvalid()

MengnanLi91 committed:
Update SolutionInvalidity.C

miaoyinb commented:
`Point` and `subdomain_id_type` added.

MengnanLi91 committed:
Update SolutionInvalidity.C

MengnanLi91 committed:
Address reviewer's comments

MengnanLi91 committed:
Update SolutionInvalidInterface.h

MengnanLi91 committed:
simplified three counters into one

MengnanLi91 committed:
Address reviewer's comment

MengnanLi91 committed:
Update SolutionInvalidity.md

zachmprince committed:
Getting tests to pass in mesh-based optimization #22995

MengnanLi91 committed:
Address reviewer's comments

MengnanLi91 committed:
Update SolutionInvalidity.md

zachmprince committed:
Addressing Daniel's comments #22995

MengnanLi91 committed:
Address reviewer's comments

MengnanLi91 committed:
Update SolutionInvalidity.md

miaoyinb commented:
@GiudGiud Let me know if any additional revision is needed before moving this to PR. Thanks.

GiudGiud commented on test/tests/meshgenerators/add_meta_data_generator/tests:
```diff
@@ -0,0 +1,223 @@
+[Tests]
+  # issues = '#11640'
+  # design = 'meshgenerators/ElementGenerator.md'
```
just those

miaoyinb replied:
Will fix this once I create the issue.

GiudGiud commented on framework/include/parser/Parser.h:
```diff
@@ -210,6 +210,18 @@ class Parser : public ConsoleStreamInterface, public hit::Walker
                                    GlobalParamsAction * global_block);
 
+  /**
+   * Template method for setting double-indexed multivalue "scalar" type parameter read from the
```
this isnt a general double indexed though. It s a vector of points.

GiudGiud commented on framework/include/parser/Parser.h:
```diff
@@ -210,6 +210,18 @@ class Parser : public ConsoleStreamInterface, public hit::Walker
 
+  /**
+   * Template method for setting double-indexed multivalue "scalar" type parameter read from the
+   * input file or command line.  Examples include "Point" and "RealVectorValue".
```
```suggestion
   * input file or command line.  Examples include several "Point"s or "RealVectorValue"s such as:
   * point_values = '0 0 0; 0 1 0; 0 0 1'
```

GiudGiud reviewed (commented):
seems almost ready to me

MengnanLi91 committed:
Update PerfGraphRegistry.h

cticenhour committed:
Address initial review (#23049)

miaoyinb commented:
@GiudGiud resolved the your last comments. If it is okay to you, I will create an issue, update the test with issue number and then submit the PR.

loganharbour committed:
Merge pull request #23005 from MengnanLi91/next_solution_invalid_22814

Improved Features for 'Add an interface to mark the solution as invalid / out-of-bounds'

GiudGiud commented:
Yeah this is great

zachmprince committed:
Merge pull request #23057 from zachmprince/optimization_gridded_data

Mesh-based parameter definition for optimization

cticenhour committed:
Merge pull request #23049 from lindsayad/sqa-doc-work-22372

Add SRS, SDD, and RTM docs for scalar transport, FSI, heat conduction modules
